File: frontend/src/types/dataset.ts
export enum DatasetStatus {
  DOWNLOADING = 'downloading',
  PROCESSING = 'processing',
  READY = 'ready',
  ERROR = 'error',
}

/**
 * Dataset schema information describing column structure.
 * Matches backend Pydantic SchemaMetadata schema.
 */
export interface SchemaMetadata {
  /** List of text/string columns in the dataset */
  text_columns: string[];
  /** Mapping of column names to their data types */
  column_info: Record<string, string>;
  /** Complete list of all columns in the dataset */
  all_columns: string[];
  /** Whether the dataset has multiple text columns */
  is_multi_column: boolean;
}

/**
 * Tokenization statistics and configuration.
 * Matches backend Pydantic TokenizationMetadata schema.
 */
export interface TokenizationMetadata {
  /** HuggingFace tokenizer identifier (e.g., 'gpt2', 'bert-base-uncased') */
  tokenizer_name: string;
  /** Name of the text column that was tokenized */
  text_column_used: string;
  /** Maximum sequence length in tokens (1-8192) */
  max_length: number;
  /** Sliding window stride (0 = no overlap) */
  stride: number;
  /** Padding strategy used (e.g., 'max_length', 'longest', 'do_not_pad') */
  padding?: string;
  /** Truncation strategy used (e.g., 'longest_first', 'only_first', 'only_second', 'do_not_truncate') */
  truncation?: string;
  /** Total number of tokens across all samples */
  num_tokens: number;
  /** Average sequence length in tokens */
  avg_seq_length: number;
  /** Minimum sequence length in tokens */
  min_seq_length: number;
  /** Maximum sequence length in tokens */
  max_seq_length: number;
  /** Median sequence length in tokens */
  median_seq_length?: number;
  /** Number of unique tokens in the tokenized dataset (vocabulary size) */
  vocab_size?: number;
  /** Distribution of sequence lengths bucketed by range (e.g., '0-100': 150) */
  length_distribution?: Record<string, number>;
  /** Distribution of samples across splits (e.g., {'train': 8000, 'validation': 1500, 'test': 500}) */
  split_distribution?: Record<string, number>;
}

/**
 * Dataset download metadata.
 * Matches backend Pydantic DownloadMetadata schema.
 */
export interface DownloadMetadata {
  /** Dataset split downloaded (e.g., 'train', 'validation', 'test') */
  split?: string;
  /** Dataset configuration name (e.g., 'en', 'zh') for multi-config datasets */
  config?: string;
  /** Whether an access token was used for download */
  access_token_provided?: boolean;
}

/**
 * Complete dataset metadata structure.
 * Matches backend Pydantic DatasetMetadata schema.
 *
 * Note: Backend uses 'schema' key but it maps to 'dataset_schema' internally.
 * Frontend receives it as 'schema' from API.
 */
export interface DatasetMetadata {
  /** Dataset schema information */
  schema?: SchemaMetadata;
  /** Tokenization statistics and configuration */
  tokenization?: TokenizationMetadata;
  /** Download metadata (split, config, etc.) */
  download?: DownloadMetadata;
}

export enum TokenizationStatus {
  QUEUED = 'queued',
  PROCESSING = 'processing',
  READY = 'ready',
  ERROR = 'error',
}

export type TokenFilterMode = 'minimal' | 'conservative' | 'standard' | 'aggressive' | 'strict';

export interface DatasetTokenization {
  id: string;
  dataset_id: string;
  model_id: string;
  max_length: number;
  tokenized_path?: string;
  tokenizer_repo_id: string;
  vocab_size?: number;
  num_tokens?: number;
  avg_seq_length?: number;
  status: TokenizationStatus | 'queued' | 'processing' | 'ready' | 'error';
  progress?: number;
  error_message?: string;
  celery_task_id?: string;
  created_at: string;
  updated_at: string;
  completed_at?: string;
  remove_all_punctuation?: boolean;
  custom_filter_chars?: string;
}

/**
 * Detailed tokenization progress information for real-time updates.
 * Extends basic progress with stage information, filtering statistics, and performance metrics.
 */
export interface DatasetTokenizationProgress {
  /** Unique identifier for the tokenization job */
  tokenization_id: string;
  /** Dataset ID this tokenization belongs to */
  dataset_id: string;
  /** Overall progress percentage (0-100) */
  progress: number;
  /** Current processing stage */
  stage: 'loading' | 'tokenizing' | 'filtering' | 'saving' | 'complete';
  /** Number of samples processed so far */
  samples_processed: number;
  /** Total number of samples to process */
  total_samples: number;
  /** Timestamp when tokenization started */
  started_at?: string;
  /** Elapsed time in seconds */
  elapsed_seconds?: number;
  /** Estimated time remaining in seconds */
  estimated_seconds_remaining?: number;
  /** Current processing rate (samples per second) */
  samples_per_second?: number;

  /** Filtering statistics (when filtering is enabled) */
  filter_stats?: {
    /** Number of samples filtered out */
    samples_filtered: number;
    /** Number of tokens classified as junk */
    junk_tokens: number;
    /** Total number of tokens processed */
    total_tokens: number;
    /** Percentage of samples filtered (0-100) */
    filter_rate: number;
  };
}

export interface DatasetTokenizationListResponse {
  data: DatasetTokenization[];
  total: number;
}

export interface Dataset {
  id: string;
  name: string;
  source: string;
  hf_repo_id?: string;
  status: DatasetStatus | 'downloading' | 'processing' | 'ready' | 'error';
  progress?: number;
  error_message?: string;
  raw_path?: string;
  num_samples?: number;
  size_bytes?: number;
  /** Structured dataset metadata with schema, tokenization, and download info */
  metadata?: DatasetMetadata;
  created_at: string;
  updated_at: string;
  /** Token filtering enabled for this dataset (applied during tokenization) */
  tokenization_filter_enabled?: boolean;
  /** Token filter mode (minimal, conservative) */
  tokenization_filter_mode?: string;
  /** Junk ratio threshold (0.0-1.0) */
  tokenization_junk_ratio_threshold?: number;
  /** List of tokenizations for this dataset */
  tokenizations?: DatasetTokenization[];
}

export interface DatasetSample {
  id: number;
  text: string;
  split: 'train' | 'validation' | 'test';
  metadata?: Record<string, any>;
  tokens?: string[];
  token_count?: number;
}

export interface DatasetStatistics {
  total_samples: number;
  total_tokens: number;
  avg_tokens_per_sample: number;
  unique_tokens: number;
  min_length: number;
  median_length: number;
  max_length: number;
  distribution?: Array<{
    range: string;
    count: number;
  }>;
}

export interface DatasetDownloadRequest {
  repo_id: string;
  access_token?: string;
  split?: string;
  config?: string;
}

export interface DatasetListResponse {
  data: Dataset[];
  pagination?: {
    page: number;
    total: number;
    has_next: boolean;
  };
}

export interface DatasetResponse {
  data: Dataset;
}

File: backend/src/schemas/dataset.py
"""
Pydantic schemas for Dataset API endpoints.

These schemas define the structure for request/response validation
and serialization for all dataset-related API operations.
"""

from datetime import datetime
from typing import Optional, Dict, Any, Union, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, field_serializer, ValidationError

from ..models.dataset import DatasetStatus
from .metadata import DatasetMetadata


class DatasetBase(BaseModel):
    """Base schema with common dataset fields."""

    name: str = Field(..., min_length=1, max_length=255, description="Dataset name")
    source: str = Field(..., min_length=1, max_length=50, description="Source type: HuggingFace, Local, or Custom")


class DatasetCreate(DatasetBase):
    """Schema for creating a new dataset."""

    hf_repo_id: Optional[str] = Field(None, max_length=255, description="HuggingFace repository ID")
    raw_path: Optional[str] = Field(None, max_length=512, description="Path to raw dataset files")
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Additional metadata")

    # Tokenization filter configuration
    tokenization_filter_enabled: bool = Field(False, description="Enable sample filtering during tokenization")
    tokenization_filter_mode: Literal["minimal", "conservative"] = Field("conservative", description="Filter mode: minimal or conservative")
    tokenization_junk_ratio_threshold: float = Field(0.7, ge=0.0, le=1.0, description="Junk ratio threshold (0.0-1.0)")


class DatasetUpdate(BaseModel):
    """Schema for updating an existing dataset."""

    name: Optional[str] = Field(None, min_length=1, max_length=255)
    status: Optional[str] = Field(None, pattern="^(downloading|ingesting|processing|ready|error)$")
    progress: Optional[float] = Field(None, ge=0, le=100)
    error_message: Optional[str] = None
    raw_path: Optional[str] = Field(None, max_length=512)
    num_samples: Optional[int] = Field(None, ge=0)
    size_bytes: Optional[int] = Field(None, ge=0)
    metadata: Optional[Union[DatasetMetadata, Dict[str, Any]]] = Field(
        None,
        description="Dataset metadata (validated structure or raw dict for backwards compatibility)"
    )
    tokenization_filter_enabled: Optional[bool] = Field(None, description="Enable sample filtering during tokenization")
    tokenization_filter_mode: Optional[Literal["minimal", "conservative", "standard", "aggressive", "strict"]] = Field(None, description="Filter mode: minimal, conservative, standard, aggressive, or strict")
    tokenization_junk_ratio_threshold: Optional[float] = Field(None, ge=0.0, le=1.0, description="Junk ratio threshold (0.0-1.0)")

    @field_validator("metadata", mode="before")
    @classmethod
    def validate_metadata(cls, v):
        """
        Validate metadata structure if provided.

        Attempts to validate against DatasetMetadata schema for structured validation.
        Falls back to raw dict for backwards compatibility with existing data.
        """
        if v is None:
            return None

        # If already a DatasetMetadata instance, return as-is
        if isinstance(v, DatasetMetadata):
            return v

        # If it's a dict, check if it looks like structured metadata
        if isinstance(v, dict):
            # Check if dict has any of the expected metadata top-level keys
            expected_keys = {"schema", "tokenization", "download"}
            has_expected_structure = any(key in v for key in expected_keys)

            if has_expected_structure:
                # Try to validate structured metadata
                try:
                    validated = DatasetMetadata(**v)
                    return validated.model_dump(by_alias=True)  # Return as dict with aliases
                except ValidationError as e:
                    # For backwards compatibility, allow raw dicts
                    # Log validation errors for debugging but don't fail
                    import logging
                    logger = logging.getLogger(__name__)
                    logger.warning(
                        f"Metadata validation failed, storing as raw dict: {e.error_count()} errors. "
                        f"First error: {e.errors()[0] if e.errors() else 'unknown'}"
                    )
                    return v
            else:
                # Unstructured metadata (no expected keys) - keep as raw dict
                return v

        return v


class DatasetResponse(DatasetBase):
    """Schema for dataset response."""

    id: UUID = Field(..., description="Unique dataset identifier")
    hf_repo_id: Optional[str] = Field(None, description="HuggingFace repository ID")
    status: str = Field(..., description="Current processing status")
    progress: Optional[float] = Field(None, description="Download/processing progress (0-100)")
    error_message: Optional[str] = Field(None, description="Error message if status is ERROR")
    raw_path: Optional[str] = Field(None, description="Path to raw dataset files")
    num_samples: Optional[int] = Field(None, description="Total number of samples in raw dataset")
    size_bytes: Optional[int] = Field(None, description="Total size in bytes")
    metadata: Dict[str, Any] = Field(
        default_factory=dict,
        description="Additional metadata",
        validation_alias="extra_metadata"  # Read from extra_metadata attribute
    )
    created_at: datetime = Field(..., description="Record creation timestamp")
    updated_at: datetime = Field(..., description="Record last update timestamp")

    # Tokenization filter configuration
    tokenization_filter_enabled: bool = Field(False, description="Enable sample filtering during tokenization")
    tokenization_filter_mode: str = Field("conservative", description="Filter mode: minimal or conservative")
    tokenization_junk_ratio_threshold: float = Field(0.7, description="Junk ratio threshold (0.0-1.0)")

    # Relationship - tokenizations for this dataset
    tokenizations: Optional[list['DatasetTokenizationResponse']] = Field(None, description="List of tokenizations for this dataset")

    @field_serializer('status')
    def serialize_status(self, status: DatasetStatus | str, _info) -> str:
        """Serialize status enum to lowercase value for frontend compatibility."""
        if isinstance(status, DatasetStatus):
            return status.value  # Returns "downloading" for frontend enum
        return str(status).lower() if status else status

    model_config = {
        "from_attributes": True,  # Enable ORM mode for SQLAlchemy models
        "populate_by_name": True,  # Allow populating by field name or alias
    }


class DatasetListResponse(BaseModel):
    """Schema for paginated list of datasets."""

    data: list[DatasetResponse] = Field(..., description="List of datasets")
    pagination: Dict[str, Any] = Field(..., description="Pagination metadata")


class DatasetDownloadRequest(BaseModel):
    """Schema for HuggingFace dataset download request."""

    repo_id: str = Field(..., min_length=1, description="HuggingFace repository ID (e.g., 'roneneldan/TinyStories')")
    access_token: Optional[str] = Field(None, description="HuggingFace access token for gated datasets")
    split: Optional[str] = Field(None, description="Dataset split to download (e.g., 'train', 'validation', 'test')")
    config: Optional[str] = Field(None, description="Dataset configuration name (e.g., 'en', 'zh') for datasets with multiple configs")

    @field_validator("repo_id")
    @classmethod
    def validate_repo_id(cls, v: str) -> str:
        """Validate HuggingFace repository ID format."""
        if "/" not in v:
            raise ValueError("repo_id must be in format 'username/dataset-name'")
        return v


class DatasetTokenizeRequest(BaseModel):
    """Schema for dataset tokenization request."""

    model_id: str = Field(..., min_length=1, description="Model ID whose tokenizer will be used for tokenization")
    max_length: int = Field(512, ge=1, le=8192, description="Maximum sequence length in tokens")
    stride: int = Field(0, ge=0, description="Sliding window stride for long sequences (0 = no overlap)")
    padding: Literal["max_length", "longest", "do_not_pad"] = Field(
        "max_length",
        description="Padding strategy: 'max_length' pads to max_length, 'longest' pads to longest in batch, 'do_not_pad' disables padding"
    )
    truncation: Literal["longest_first", "only_first", "only_second", "do_not_truncate"] = Field(
        "longest_first",
        description="Truncation strategy: 'longest_first' truncates longest sequence first, 'only_first' truncates only first sequence, 'only_second' truncates only second sequence, 'do_not_truncate' disables truncation"
    )
    add_special_tokens: bool = Field(
        True,
        description="Add special tokens (BOS, EOS, PAD, etc.) - Recommended for most models"
    )
    return_attention_mask: bool = Field(
        True,
        description="Return attention mask - Set to False to save memory if model doesn't use attention masks"
    )
    enable_cleaning: bool = Field(
        True,
        description="Enable text cleaning (removes HTML tags, control characters, excessive punctuation, normalizes Unicode) - Recommended for better feature quality"
    )
    tokenization_filter_enabled: bool = Field(
        False,
        description="Enable sample filtering during tokenization (Stage 1 filtering) - Filters out low-quality samples based on junk ratio threshold"
    )
    tokenization_filter_mode: Literal["minimal", "conservative", "standard", "aggressive", "strict"] = Field(
        "conservative",
        description="Filter mode: 'minimal' (only control chars), 'conservative' (+ whitespace), 'standard' (+ punctuation), 'aggressive' (+ short tokens), 'strict' (+ ALL punctuation)"
    )
    tokenization_junk_ratio_threshold: float = Field(
        0.7,
        ge=0.0,
        le=1.0,
        description="Junk ratio threshold for filtering samples (0.0-1.0). Samples with junk ratio above this threshold are filtered out. Higher values = more permissive."
    )
    remove_all_punctuation: bool = Field(
        False,
        description="Remove ALL punctuation characters from tokens (overrides filter_mode setting)"
    )
    custom_filter_chars: Optional[str] = Field(
        None,
        max_length=255,
        description="Custom characters to filter from tokens (e.g., '~@#$%')"
    )

    @field_validator("stride")
    @classmethod
    def validate_stride(cls, v: int, info) -> int:
        """Validate that stride is less than or equal to max_length."""
        max_length = info.data.get("max_length", 512)
        if v > max_length:
            raise ValueError(f"stride ({v}) must be less than or equal to max_length ({max_length})")
        return v


class TokenizePreviewRequest(BaseModel):
    """Schema for tokenization preview request."""

    tokenizer_name: str = Field(..., min_length=1, description="HuggingFace tokenizer name")
    text: str = Field(..., min_length=1, max_length=1000, description="Text to tokenize (max 1000 chars)")
    max_length: int = Field(512, ge=1, le=8192, description="Maximum sequence length")
    padding: Literal["max_length", "longest", "do_not_pad"] = Field("max_length", description="Padding strategy")
    truncation: Literal["longest_first", "only_first", "only_second", "do_not_truncate"] = Field("longest_first", description="Truncation strategy")
    add_special_tokens: bool = Field(True, description="Add special tokens (BOS, EOS, etc.)")
    return_attention_mask: bool = Field(True, description="Return attention mask")


class TokenInfo(BaseModel):
    """Information about a single token."""

    id: int = Field(..., description="Token ID")
    text: str = Field(..., description="Token text")
    type: Literal["special", "regular"] = Field(..., description="Token type")
    position: int = Field(..., ge=0, description="Position in sequence")


class TokenizePreviewResponse(BaseModel):
    """Schema for tokenization preview response."""

    tokens: list[TokenInfo] = Field(..., description="List of tokens with metadata")
    attention_mask: Optional[list[int]] = Field(None, description="Attention mask (if requested)")
    token_count: int = Field(..., ge=0, description="Total number of tokens")
    sequence_length: int = Field(..., ge=0, description="Length of tokenized sequence")
    special_token_count: int = Field(..., ge=0, description="Number of special tokens")


class DatasetTokenizationResponse(BaseModel):
    """Schema for dataset tokenization response."""

    id: str = Field(..., description="Unique tokenization identifier")
    dataset_id: UUID = Field(..., description="Parent dataset ID")
    model_id: str = Field(..., description="Model whose tokenizer was used")
    tokenized_path: Optional[str] = Field(None, description="Path to tokenized dataset files")
    tokenizer_repo_id: str = Field(..., description="HuggingFace tokenizer repository ID")
    vocab_size: Optional[int] = Field(None, description="Vocabulary size for this tokenization")
    num_tokens: Optional[int] = Field(None, description="Total number of tokens in tokenized dataset")
    avg_seq_length: Optional[float] = Field(None, description="Average sequence length in tokens")
    status: str = Field(..., description="Current tokenization status")
    progress: Optional[float] = Field(None, description="Tokenization progress (0-100)")
    error_message: Optional[str] = Field(None, description="Error message if status is ERROR")
    celery_task_id: Optional[str] = Field(None, description="Celery task ID for async tokenization")
    created_at: datetime = Field(..., description="Record creation timestamp")
    updated_at: datetime = Field(..., description="Record last update timestamp")
    completed_at: Optional[datetime] = Field(None, description="Timestamp when tokenization completed")
    remove_all_punctuation: bool = Field(False, description="If true, removes ALL punctuation characters from tokens")
    custom_filter_chars: Optional[str] = Field(None, description="Custom characters to filter (e.g., '~@#$%')")

    model_config = {
        "from_attributes": True,  # Enable ORM mode for SQLAlchemy models
    }


class DatasetTokenizationListResponse(BaseModel):
    """Schema for list of dataset tokenizations."""

    data: list[DatasetTokenizationResponse] = Field(..., description="List of tokenizations")
    total: int = Field(..., ge=0, description="Total number of tokenizations")
